controllers/remote/cluster_cache.go

```diff
@@ -54,319 +48,206 @@ const (
-	watches     map[client.ObjectKey]map[watchInfo]struct{}
+	lock      sync.RWMutex
+	accessors map[client.ObjectKey]*accessor
 }
 
 // NewClusterCacheTracker creates a new ClusterCacheTracker.
 func NewClusterCacheTracker(log logr.Logger, manager ctrl.Manager) (*ClusterCacheTracker, error) {
-	m := &ClusterCacheTracker{
-		log:               log,
-		client:            manager.GetClient(),
-		scheme:            manager.GetScheme(),
-		delegatingClients: make(map[client.ObjectKey]*client.DelegatingClient),
-		clusterCaches:     make(map[client.ObjectKey]*clusterCache),
-		watches:           make(map[client.ObjectKey]map[watchInfo]struct{}),
-	}
-
-	return m, nil
+	return &ClusterCacheTracker{
+		log:       log,
+		client:    manager.GetClient(),
+		scheme:    manager.GetScheme(),
+		accessors: make(map[client.ObjectKey]*accessor),
+	}, nil
 }
 
-// Watcher is a scoped-down interface from Controller that only knows how to watch.
-type Watcher interface {
-	// Watch watches src for changes, sending events to eventHandler if they pass predicates.
-	Watch(src source.Source, eventHandler handler.EventHandler, predicates ...predicate.Predicate) error
-}
+// GetClient returns a client for the given cluster.
+func (t *ClusterCacheTracker) GetClient(ctx context.Context, cluster client.ObjectKey) (client.Client, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
 
-// watchInfo is used as a map key to uniquely identify a watch. Because predicates is a slice, it cannot be included.
-type watchInfo struct {
-	watcher Watcher
-	gvk     schema.GroupVersionKind
-
-	// Comparing the eventHandler as an interface doesn't work because reflect.DeepEqual
-	// will assert functions are false if they are non-nil.
-	// Use a signature string representation instead as this can be compared.
-	// The signature function is expected to produce a unique output for each unique handler
-	// function that is passed to it.
-	// In combination with the watcher, this should be enough to identify unique watches.
-	eventHandlerSignature string
-}
+	accessor, err := t.getAccessorLH(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
 
-// eventHandlerSignature generates a unique identifier for the given eventHandler by
-// printing it to a string using "%#v".
-// Eg "&handler.EnqueueRequestsFromMapFunc{ToRequests:(handler.ToRequestsFunc)(0x271afb0)}"
-func eventHandlerSignature(h handler.EventHandler) string {
-	return fmt.Sprintf("%#v", h)
+	return accessor.client, nil
 }
 
-// watchExists returns true if watch has already been established. This does NOT hold any lock.
-func (m *ClusterCacheTracker) watchExists(cluster client.ObjectKey, watch watchInfo) bool {
-	watchesForCluster, clusterFound := m.watches[cluster]
-	if !clusterFound {
-		return false
-	}
-
-	for w := range watchesForCluster {
-		if reflect.DeepEqual(w, watch) {
-			return true
-		}
-	}
-	return false
+// accessor represents the combination of a client, cache, and watches for a remote cluster.
+type accessor struct {
```

Comment:
Might be 
```suggestion
type remoteClusterAccessor struct {
```
or 
```suggestion
type clusterAccessor struct {
```
to be more specific about the intent of this struct
if this is ok, also func names below should be changed accordingly

Reply:
I'll change it to `clusterAccessor`.